packages/flutter/lib/src/widgets/form.dart

```diff
@@ -454,6 +452,7 @@ class FormField<T> extends StatefulWidget {
     super.key,
     required this.builder,
     this.onSaved,
+    this.forceErrorText,
     this.validator,
```

Comment:
Is it true that both of these should not be used at the same time? Or what kind of cases do you expect like that?

Reply:
When using forceErrorText to display an error message, running the validator function to check the form's validity becomes redundant, as illustrated in the following code:
```dart
 void _validate() {
    if (widget.forceErrorText != null) {
      _errorText.value = widget.forceErrorText;
      // Skip validating if error is forced.
      return;
    }
    if (widget.validator != null) {
      _errorText.value = widget.validator!(_value);
    } else {
      _errorText.value = null;
    }
  }
```
The validator function provided by the user will not be called, and _validate will return false unless forceErrorText is set to null.

The initial issue was that InputDecoration has an errorText parameter, but it only affects the decoration, not the FormState. This results in a visual error state without affecting the form validation, causing _validate to return true.

To address this, I added a forceErrorText parameter to apply the error to both the decoration and the form state. This ensures that the form behaves as expected when an error is forced. The term "force" was chosen to indicate that the validator will not be called in this scenario.

Reply:
Then I think it would make sense to disallow `validator` when `forceErrorText` is passed. I think you should add an `assert` in the initializers here to enforce that.

Reply:
The problem is that we need to use both. The main idea is that we cannot make the validator function run asynchronously because the form must always have a synchronous state (valid/invalid). Handling asynchronous validation errors within the form widget would introduce unnecessary complexity.

Instead, `forceErrorText` was introduced so the user can run the normal `validator` function to check for any potential errors, then perform any asynchronous computations. If an async validation error occurs, the user can simply set `forceErrorText` to display the error and make the `formState.validate()` function return false, so the form state is invalid. This approach keeps the form state management simple and synchronous.

Here, the return of the `validator` function becomes redundant as the user is knowingly forcing the error — hence the name forceErrorText — which is why there is no point in allowing the `validator` to be called. Adding an assertion to limit the use of only one of them would burden the user to alternate between providing the `forceErrorText` and a `validator` function. I think they can work together perfectly, as you don't have to call `formState.validate` to show the error of the forceErrorText. The moment the TextFormField is built with a forceErrorText, the error text will be displayed immediately.

The user is responsible for setting `forceErrorText` to `null` when the text value changes. This task should not be handled by the FormState widget, as it is beyond its intended scope.

Reply:
Would you mind writing an example of that pattern you're describing? It could be something like https://github.com/flutter/flutter/blob/master/examples/api/lib/material/text_form_field/text_form_field.1.dart if you're up for it. Or if that's overkill then in a test or just paste the code here. I want to see how you intend to use this so I can have a better understanding of the API.

Reply:
That example was great and helped me clearly see your point. I think you're right that it should be possible to pass both parameters at the same time.

We should probably document this behavior where we can, though. Maybe on the docs for each `forceErrorText` and `validator` parameter you should mention the other parameter and explain that `forceErrorText` will override `validator`.